### PR TITLE
Change Dispatch.h to use ScalarType over Type

### PR DESCRIPTION
**Pull request by li-roy**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#17527 Change Dispatch.h to use ScalarType over Type**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14235395/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17529 Remove some simple use cases of Type::ScalarType()&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14237932/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17530 Small clean up of aten_op&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14237931/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17601 Store ScalarType and Backend instead of Type in TensorIterator&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14274754/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17785 Remove Type::elementSizeInBytes&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14379074/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17723 Store python default type as PyTensorType instead of at::Type&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14351082/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17786 Pass ScalarType separately from Type in python constructors&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14379075/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17792 Remove Type::ScalarType()&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14381774/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17603 Remove conversion operator from Type to TensorOptions&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14276588/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17787 Add ScalarType arg to Type::options()&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14379076/)

Differential Revision: [D14235395](https://our.internmc.facebook.com/intern/diff/D14235395/)

### CODE REVIEW

**Review by gchanan — CHANGES_REQUESTED** [review]
This made a number of error messages worse.

For example, let's say I have a CPU and CUDA implementation of the same function, fn, which I give in both cases to the dispatch macro as "fn".  And as with most functions, I implement the half variant on CUDA, but not CPU.

Then, the user tries to call "fn" with CPU half, and gets the message "fn is not implemented for Half" which is wrong.

Can you just go through and find the cases where this exists (if such a case exists, note it _will_ exist if you shepherd TH functions through this macro).

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -474,7 +474,7 @@ Tensor& tril_cpu_(Tensor &self, int64_t k) {
   bool inplace = checkTrilTriuBatchContiguous(self);
   Tensor self_c = inplace ? self : self.contiguous();
   Tensor result = inplace ? self : at::empty_like(self);
-  AT_DISPATCH_ALL_TYPES(self.type(), "tril", [&]{
+  AT_DISPATCH_ALL_TYPES(self.scalar_type(), "tril", [&]{
```
Comment (ezyang) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp]: `tril_cpu` here

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -489,7 +489,7 @@ Tensor& tril_cpu_out(Tensor &result, const Tensor& self, int64_t k) {
     return result;
   }
   Tensor self_c = checkTrilTriuBatchContiguous(self) ? self : self.contiguous();
-  AT_DISPATCH_ALL_TYPES(self.type(), "tril", [&]{
+  AT_DISPATCH_ALL_TYPES(self.scalar_type(), "tril", [&]{
```
Comment (ezyang) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp]: ditto

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 3]
```diff
@@ -508,7 +508,7 @@ Tensor& triu_cpu_(Tensor &self, int64_t k) {
   bool inplace = checkTrilTriuBatchContiguous(self);
   Tensor self_c = inplace ? self : self.contiguous();
   Tensor result = inplace ? self : at::empty_like(self);
-  AT_DISPATCH_ALL_TYPES(self.type(), "triu", [&]{
+  AT_DISPATCH_ALL_TYPES(self.scalar_type(), "triu", [&]{
```
Comment (ezyang) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp]: ditto

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 4]
```diff
@@ -523,7 +523,7 @@ Tensor& triu_cpu_out(Tensor &result, const Tensor& self, int64_t k) {
     return result;
   }
   Tensor self_c = checkTrilTriuBatchContiguous(self) ? self : self.contiguous();
-  AT_DISPATCH_ALL_TYPES(self.type(), "triu", [&]{
+  AT_DISPATCH_ALL_TYPES(self.scalar_type(), "triu", [&]{
```
Comment (ezyang) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp]: ditto

File: aten/src/ATen/native/Copy.cpp [hunk 5]
```diff
@@ -115,7 +115,7 @@ void _copy_same_type__cpu(Tensor& self, const Tensor& src) {
 #ifdef _OPENMP
       if (!in_parallel_region()) {
         AT_DISPATCH_ALL_TYPES_AND(
-          at::ScalarType::Half, self.type(), "_copy_same_type_", [&]() {
+          at::ScalarType::Half, self.scalar_type(), "_copy_same_type_", [&]() {
```
Comment (ezyang) [inline aten/src/ATen/native/Copy.cpp]: cpu

File: aten/src/ATen/native/Copy.cpp [hunk 6]
```diff
@@ -134,7 +134,7 @@ void _copy_same_type__cpu(Tensor& self, const Tensor& src) {
 
   if (serial_path) {
     AT_DISPATCH_ALL_TYPES_AND(
-      at::ScalarType::Half, self.type(), "_copy_same_type_", [&]() {
+      at::ScalarType::Half, self.scalar_type(), "_copy_same_type_", [&]() {
```
Comment (ezyang) [inline aten/src/ATen/native/Copy.cpp]: cpu

File: aten/src/ATen/native/FractionalMaxPool2d.cpp [hunk 7]
```diff
@@ -178,7 +178,7 @@ void fractional_max_pool2d_out_cpu_template(
     indices.resize_({numBatch, numPlanes, outputH, outputW});
   }
 
-  AT_DISPATCH_FLOATING_TYPES(input.type(),
+  AT_DISPATCH_FLOATING_TYPES(input.scalar_type(),
   "fractional_max_pool2d_out_frame", [&] {
```
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool2d.cpp]: cpu

File: aten/src/ATen/native/FractionalMaxPool2d.cpp [hunk 8]
```diff
@@ -295,7 +295,7 @@ Tensor& fractional_max_pool2d_backward_out_cpu_template(
 
   /* backprop */
   AT_DISPATCH_FLOATING_TYPES(
-    input.type(), "fractional_max_pool2d_backward_out_frame", [&] {
+    input.scalar_type(), "fractional_max_pool2d_backward_out_frame", [&] {
```
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool2d.cpp]: cpu

File: aten/src/ATen/native/FractionalMaxPool3d.cpp [hunk 9]
```diff
@@ -201,7 +201,7 @@ void fractional_max_pool3d_out_cpu_template(
     indices.resize_({numBatch, numPlanes, outputT, outputH, outputW});
   }
   AT_DISPATCH_FLOATING_TYPES(
-    input.type(),
+    input.scalar_type(),
     "fractional_max_pool3d_out_frame",
```
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool3d.cpp]: cpu

**Review by ezyang — CHANGES_REQUESTED** [review]
I stopped something like 30% of the way through the patch, but it seems there are still a lot of missing sites

**Comment by li-roy** [comment]
@ezyang I only changed the strings that are used on both cpu and gpu kernels.

**Review by gchanan — APPROVED** [review]

**Pull request closed without merging**